File: source-connector/src/main/java/sourceconnector/config/WorkerConfiguration.java
package sourceconnector.config;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import sourceconnector.domain.connect.FileTaskAssignor;
import sourceconnector.domain.connect.OffsetRecordService;
import sourceconnector.domain.connect.TaskAssignor;
import sourceconnector.domain.connect.Worker;
import offsetmanager.domain.file.FileKey;
import sourceconnector.repository.file.FileLister;

import java.io.IOException;

@Configuration
public class WorkerConfiguration {

  @Bean
  public TaskAssignor taskAssignor(
    FileLister fileLister,
    StorageConfig storageConfig,
    FileSearchConfigs fileSearchConfigs,
    ConnectConfig connectConfig,
    OffsetRecordService offsetRecordService
  ) throws IOException {

    FileKey[] inputFileKeys = storageConfig.getAllFileKeys().toArray(new FileKey[0]);

    if (fileSearchConfigs.isRecursive()) {
      var foundFileKeys = fileLister.listFilesRecursively(inputFileKeys);
      return new FileTaskAssignor(foundFileKeys, connectConfig.taskCount(), offsetRecordService);
    } else {
      var foundFileKeys = fileLister.listFiles(inputFileKeys);
      return new FileTaskAssignor(foundFileKeys, connectConfig.taskCount(), offsetRecordService);
    }
  }

  @Bean
  public Worker worker(
    TaskAssignor taskAssignor,
    @Value("${JOB_INDEX}")
    int jobIndex
  ) {
    return new Worker(jobIndex, taskAssignor);
  }
}

File: source-connector/src/main/kotlin/sourceconnector/config/WorkerConfiguration.kt
package sourceconnector.config

import offsetmanager.domain.file.FileKey
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import sourceconnector.domain.connect.FileTaskAssignor
import sourceconnector.domain.connect.OffsetRecordService
import sourceconnector.domain.connect.TaskAssignor
import sourceconnector.domain.connect.Worker
import sourceconnector.repository.file.FileLister
import java.io.IOException

@Configuration
class WorkerConfiguration {
  @Bean
  @Throws(IOException::class)
  fun taskAssignor(
    fileLister: FileLister,
    storageConfig: StorageConfig,
    fileSearchConfigs: FileSearchConfigs,
    connectConfig: ConnectConfig,
    offsetRecordService: OffsetRecordService
  ): TaskAssignor {
    val inputFileKeys = storageConfig.allFileKeys.toTypedArray<FileKey>()

    if (fileSearchConfigs.isRecursive) {
      val foundFileKeys = fileLister.listFilesRecursively(*inputFileKeys)
      return FileTaskAssignor(foundFileKeys, connectConfig.taskCount, offsetRecordService)
    } else {
      val foundFileKeys = fileLister.listFiles(*inputFileKeys)
      return FileTaskAssignor(foundFileKeys, connectConfig.taskCount, offsetRecordService)
    }
  }

  @Bean
  fun worker(
    taskAssignor: TaskAssignor,
    @Value("\${JOB_INDEX}") jobIndex: Int
  ): Worker {
    return Worker(jobIndex, taskAssignor)
  }
}
